diff --git a/arch/risc-v/Kconfig b/arch/risc-v/Kconfig
index 5a78e4c65a..01094bb145 100644
--- a/arch/risc-v/Kconfig
+++ b/arch/risc-v/Kconfig
@@ -99,9 +99,11 @@ config ARCH_CHIP_ESP32C3_GENERIC
 	select ARCH_RV_ISA_C
 	select ARCH_VECNOTIRQ
 	select ARCH_HAVE_BOOTLOADER if !ESPRESSIF_SIMPLE_BOOT
+	select ARCH_HAVE_ELF_EXECUTABLE
 	select ARCH_HAVE_MPU
 	select ARCH_HAVE_RESET
 	select ARCH_HAVE_RNG
+	select ARCH_HAVE_TEXT_HEAP
 	select ARCH_HAVE_TICKLESS
 	select LIBC_ARCH_ATOMIC
 	select LIBC_ARCH_MEMCPY
diff --git a/arch/risc-v/src/cmake/Toolchain.cmake b/arch/risc-v/src/cmake/Toolchain.cmake
index fb0e9d1db5..9cf90a0cda 100644
--- a/arch/risc-v/src/cmake/Toolchain.cmake
+++ b/arch/risc-v/src/cmake/Toolchain.cmake
@@ -116,11 +116,7 @@ set(CMAKE_ASM_ARCHIVE_CREATE "<CMAKE_AR> rcs <TARGET> <LINK_FLAGS> <OBJECTS>")
 if(CONFIG_DEBUG_CUSTOMOPT)
   add_compile_options(${CONFIG_DEBUG_OPTLEVEL})
 elseif(CONFIG_DEBUG_FULLOPT)
-  if(CONFIG_ARCH_TOOLCHAIN_CLANG)
-    add_compile_options(-Oz)
-  else()
-    add_compile_options(-Os)
-  endif()
+  add_compile_options(-Oz)
 endif()
 
 if(NOT CONFIG_DEBUG_NOOPT)
diff --git a/arch/risc-v/src/common/espressif/Wireless.mk b/arch/risc-v/src/common/espressif/Wireless.mk
index e2e627ca9f..54b7e0874e 100644
--- a/arch/risc-v/src/common/espressif/Wireless.mk
+++ b/arch/risc-v/src/common/espressif/Wireless.mk
@@ -28,7 +28,7 @@ EXTRA_LIBPATHS += -L $(ARCH_SRCDIR)$(DELIM)chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$
 EXTRA_LIBPATHS += -L $(ARCH_SRCDIR)$(DELIM)chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$(DELIM)components$(DELIM)esp_phy$(DELIM)lib$(DELIM)$(CHIP_SERIES)
 EXTRA_LIBPATHS += -L $(ARCH_SRCDIR)$(DELIM)chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$(DELIM)components$(DELIM)esp_wifi$(DELIM)lib$(DELIM)$(CHIP_SERIES)
 
-EXTRA_LIBS += -lphy -lcoexist -lmesh
+EXTRA_LIBS += -lphy -lcoexist
 
 ifeq ($(CONFIG_ESPRESSIF_WIFI),y)
 
@@ -45,6 +45,35 @@ INCLUDES += $(INCDIR_PREFIX)$(ARCH_SRCDIR)$(DELIM)chip$(DELIM)$(ESP_HAL_3RDPARTY
 INCLUDES += $(INCDIR_PREFIX)$(ARCH_SRCDIR)$(DELIM)chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$(DELIM)components$(DELIM)mbedtls$(DELIM)port$(DELIM)include
 INCLUDES += $(INCDIR_PREFIX)$(ARCH_SRCDIR)$(DELIM)chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$(DELIM)nuttx$(DELIM)include$(DELIM)mbedtls
 
+CHIP_CSRCS += net_sockets.c
+CHIP_CSRCS += pkcs12.c
+CHIP_CSRCS += ssl_ciphersuites.c
+CHIP_CSRCS += ssl_client.c
+CHIP_CSRCS += ssl_msg.c
+CHIP_CSRCS += ssl_tls.c
+CHIP_CSRCS += ssl_tls12_client.c
+CHIP_CSRCS += ssl_tls12_server.c
+CHIP_CSRCS += x509.c
+CHIP_CSRCS += x509_crt.c
+CHIP_CSRCS += mbedtls_sslutils.c
+
+VPATH += chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$(DELIM)components$(DELIM)hal
+VPATH += chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$(DELIM)components$(DELIM)mbedtls$(DELIM)port$(DELIM)aes
+VPATH += chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$(DELIM)components$(DELIM)mbedtls$(DELIM)port$(DELIM)aes$(DELIM)block
+VPATH += chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$(DELIM)components$(DELIM)mbedtls$(DELIM)port$(DELIM)sha$(DELIM)block
+
+INCLUDES += $(INCDIR_PREFIX)$(ARCH_SRCDIR)$(DELIM)chip$(DELIM)$(ESP_HAL_3RDPARTY_REPO)$(DELIM)components$(DELIM)mbedtls$(DELIM)port$(DELIM)aes$(DELIM)dma$(DELIM)include
+
+CHIP_CSRCS += aes_hal.c
+CHIP_CSRCS += sha_hal.c
+CHIP_CSRCS += esp_aes.c
+CHIP_CSRCS += esp_aes_common.c
+CHIP_CSRCS += esp_aes_gcm.c
+CHIP_CSRCS += esp_aes_xts.c
+CHIP_CSRCS += esp_sha1.c
+CHIP_CSRCS += esp_sha256.c
+CHIP_CSRCS += sha.c
+
 ### Define Espressif's configs for mbedTLS
 
 CFLAGS += $(DEFINE_PREFIX)MBEDTLS_CONFIG_FILE="<mbedtls/esp_config.h>"
diff --git a/arch/risc-v/src/common/espressif/esp_wifi_utils.h b/arch/risc-v/src/common/espressif/esp_wifi_utils.h
index 738fbbed08..17914d9554 100644
--- a/arch/risc-v/src/common/espressif/esp_wifi_utils.h
+++ b/arch/risc-v/src/common/espressif/esp_wifi_utils.h
@@ -45,6 +45,8 @@ extern "C"
  * Public Function Prototypes
  ****************************************************************************/
 
+struct iwreq;
+
 /****************************************************************************
  * Name: esp_wifi_start_scan
  *
diff --git a/arch/risc-v/src/common/riscv_initialize.c b/arch/risc-v/src/common/riscv_initialize.c
index 9d0f1c81d7..6305a4a565 100644
--- a/arch/risc-v/src/common/riscv_initialize.c
+++ b/arch/risc-v/src/common/riscv_initialize.c
@@ -130,3 +130,8 @@ void up_initialize(void)
 
   board_autoled_on(LED_IRQSENABLED);
 }
+
+uint32_t mesh_sta_auth_expire_time(void)
+{
+    return 0;
+}
diff --git a/arch/risc-v/src/esp32c3/Make.defs b/arch/risc-v/src/esp32c3/Make.defs
index 97e5963f4b..e5b1a6e2c1 100644
--- a/arch/risc-v/src/esp32c3/Make.defs
+++ b/arch/risc-v/src/esp32c3/Make.defs
@@ -28,4 +28,8 @@ CHIP_CSRCS += esp_coex_adapter.c esp_wifi_adapter.c
 EXTRA_LIBS += -lcore -lnet80211 -lpp
 endif
 
+ifeq ($(CONFIG_ARCH_USE_TEXT_HEAP),y)
+CHIP_CSRCS += esp32c3_textheap.c
+endif
+
 CFLAGS += ${DEFINE_PREFIX}_RETARGETABLE_LOCKING
diff --git a/binfmt/libelf/libelf_symbols.c b/binfmt/libelf/libelf_symbols.c
index bbc232cc76..1d9c1fe3a9 100644
--- a/binfmt/libelf/libelf_symbols.c
+++ b/binfmt/libelf/libelf_symbols.c
@@ -304,6 +304,8 @@ int elf_symvalue(FAR struct elf_loadinfo_s *loadinfo, FAR Elf_Sym *sym,
           {
             berr("SHN_UNDEF: Exported symbol \"%s\" not found\n",
                  loadinfo->iobuffer);
+            printf("SHN_UNDEF: Exported symbol \"%s\" not found\n",
+                 loadinfo->iobuffer);
             return -ENOENT;
           }
 
diff --git a/boards/risc-v/esp32c3/common/src/esp_board_spiflash.c b/boards/risc-v/esp32c3/common/src/esp_board_spiflash.c
index 7e1c40ccb7..7ee8150239 100644
--- a/boards/risc-v/esp32c3/common/src/esp_board_spiflash.c
+++ b/boards/risc-v/esp32c3/common/src/esp_board_spiflash.c
@@ -58,18 +58,105 @@
  * Private Types
  ****************************************************************************/
 
+#ifdef CONFIG_ESPRESSIF_HAVE_OTA_PARTITION
+
+struct ota_partition_s
+{
+  uint32_t    offset;          /* Partition offset from the beginning of MTD */
+  uint32_t    size;            /* Partition size in bytes */
+  const char *devpath;         /* Partition device path */
+};
+
+#endif
+
 /****************************************************************************
  * Private Function Prototypes
  ****************************************************************************/
 
+#ifdef CONFIG_ESPRESSIF_HAVE_OTA_PARTITION
+static int init_ota_partitions(void);
+#endif
+
 /****************************************************************************
  * Private Data
  ****************************************************************************/
 
+#ifdef CONFIG_ESPRESSIF_HAVE_OTA_PARTITION
+static const struct ota_partition_s g_ota_partition_table[] =
+{
+  {
+    .offset  = CONFIG_ESPRESSIF_OTA_PRIMARY_SLOT_OFFSET,
+    .size    = CONFIG_ESPRESSIF_OTA_SLOT_SIZE,
+    .devpath = CONFIG_ESPRESSIF_OTA_PRIMARY_SLOT_DEVPATH
+  },
+  {
+    .offset  = CONFIG_ESPRESSIF_OTA_SECONDARY_SLOT_OFFSET,
+    .size    = CONFIG_ESPRESSIF_OTA_SLOT_SIZE,
+    .devpath = CONFIG_ESPRESSIF_OTA_SECONDARY_SLOT_DEVPATH
+  },
+  {
+    .offset  = CONFIG_ESPRESSIF_OTA_SCRATCH_OFFSET,
+    .size    = CONFIG_ESPRESSIF_OTA_SCRATCH_SIZE,
+    .devpath = CONFIG_ESPRESSIF_OTA_SCRATCH_DEVPATH
+  }
+};
+#endif
+
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: init_ota_partitions
+ *
+ * Description:
+ *   Initialize partitions that are dedicated to firmware OTA update.
+ *
+ * Input Parameters:
+ *   None.
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ESPRESSIF_HAVE_OTA_PARTITION
+static int init_ota_partitions(void)
+{
+  struct mtd_dev_s *mtd;
+#ifdef CONFIG_BCH
+  char blockdev[18];
+#endif
+  int ret = OK;
+
+  for (int i = 0; i < nitems(g_ota_partition_table); ++i)
+    {
+      const struct ota_partition_s *part = &g_ota_partition_table[i];
+      mtd = esp_spiflash_alloc_mtdpart(part->offset, part->size);
+
+      ret = ftl_initialize(i, mtd);
+      if (ret < 0)
+        {
+          ferr("ERROR: Failed to initialize the FTL layer: %d\n", ret);
+          return ret;
+        }
+
+#ifdef CONFIG_BCH
+      snprintf(blockdev, sizeof(blockdev), "/dev/mtdblock%d", i);
+
+      ret = bchdev_register(blockdev, part->devpath, false);
+      if (ret < 0)
+        {
+          ferr("ERROR: bchdev_register %s failed: %d\n", part->devpath, ret);
+          return ret;
+        }
+#endif
+    }
+
+  return ret;
+}
+#endif
+
 /****************************************************************************
  * Name: setup_smartfs
  *
@@ -406,6 +493,14 @@ int board_spiflash_init(void)
 
   esp_spiflash_init();
 
+#ifdef CONFIG_ESPRESSIF_HAVE_OTA_PARTITION
+  ret = init_ota_partitions();
+  if (ret < 0)
+    {
+      return ret;
+    }
+#endif
+
   ret = init_storage_partition();
   if (ret < 0)
     {
diff --git a/boards/risc-v/esp32c3/esp32c3-generic/src/Make.defs b/boards/risc-v/esp32c3/esp32c3-generic/src/Make.defs
index b23cab8ea0..500ed940a5 100644
--- a/boards/risc-v/esp32c3/esp32c3-generic/src/Make.defs
+++ b/boards/risc-v/esp32c3/esp32c3-generic/src/Make.defs
@@ -38,6 +38,10 @@ ifeq ($(CONFIG_ARCH_BUTTONS),y)
   CSRCS += esp32c3_buttons.c
 endif
 
+ifeq ($(CONFIG_ETC_ROMFS),y)
+  RCRAWS = etc/init.d/rcS
+endif
+
 DEPPATH += --dep-path board
 VPATH += :board
 CFLAGS += ${INCDIR_PREFIX}$(TOPDIR)$(DELIM)arch$(DELIM)$(CONFIG_ARCH)$(DELIM)src$(DELIM)board$(DELIM)board
diff --git a/libs/libc/libc.csv b/libs/libc/libc.csv
index 3623e7ff62..04f50a06e6 100644
--- a/libs/libc/libc.csv
+++ b/libs/libc/libc.csv
@@ -212,6 +212,7 @@
 "pthread_condattr_init","pthread.h","!defined(CONFIG_DISABLE_PTHREAD)","int","FAR pthread_condattr_t *"
 "pthread_condattr_setclock","pthread.h","!defined(CONFIG_DISABLE_PTHREAD)","int","FAR pthread_condattr_t *","clockid_t"
 "pthread_create","pthread.h","!defined(CONFIG_DISABLE_PTHREAD)","int","FAR pthread_t *","FAR const pthread_attr_t *","pthread_startroutine_t","pthread_addr_t"
+"pthread_exit","pthread.h","!defined(CONFIG_DISABLE_PTHREAD)","noreturn","pthread_addr_t"
 "pthread_getname_np","pthread.h","!defined(CONFIG_DISABLE_PTHREAD)","int","pthread_t","char *","size_t"
 "pthread_getspecific","pthread.h","!defined(CONFIG_DISABLE_PTHREAD) && CONFIG_TLS_NELEM > 0","FAR void *","pthread_key_t"
 "pthread_key_create","pthread.h","!defined(CONFIG_DISABLE_PTHREAD) && CONFIG_TLS_NELEM > 0","int","FAR pthread_key_t *","FAR void (*) (void *)|FAR void *"
@@ -248,8 +249,13 @@
 "scandir","dirent.h","","int","FAR const char *","FAR struct dirent ***","FAR int (*)(const struct dirent *)|FAR void *","FAR void *"
 "sched_get_priority_max","sched.h","","int","int"
 "sched_get_priority_min","sched.h","","int","int"
+"sem_destroy","semaphore.h","","int","FAR sem_t *"
 "sem_getvalue","semaphore.h","","int","FAR sem_t *","FAR int *"
 "sem_init","semaphore.h","","int","FAR sem_t *","int","unsigned int"
+"sem_post","semaphore.h","","int","FAR sem_t *"
+"sem_timedwait","semaphore.h","","int","FAR sem_t *","FAR const struct timespec *"
+"sem_trywait","semaphore.h","","int","FAR sem_t *"
+"sem_wait","semaphore.h","","int","FAR sem_t *"
 "setlocale","locale.h","defined(CONFIG_LIBC_LOCALE)","FAR char *","int","FAR const char *"
 "setlogmask","syslog.h","","int","int"
 "setpriority","sys/resource.h","","int","int","id_t","int"
diff --git a/libs/libc/machine/risc-v/arch_elf.c b/libs/libc/machine/risc-v/arch_elf.c
index 8e826bddc5..ba43a6617b 100644
--- a/libs/libc/machine/risc-v/arch_elf.c
+++ b/libs/libc/machine/risc-v/arch_elf.c
@@ -78,6 +78,7 @@ static struct rname_code_s _rname_table[] =
   {"CALL_PLT", R_RISCV_CALL_PLT},
   {"BRANCH", R_RISCV_BRANCH},
   {"JAL", R_RISCV_JAL},
+  {"ALIGN", R_RISCV_ALIGN},
   {"RVC_JUMP", R_RISCV_RVC_JUMP},
   {"RVC_BRANCH", R_RISCV_RVC_BRANCH},
   {"32_PCREL", R_RISCV_32_PCREL},
@@ -104,6 +105,11 @@ static const char *_get_rname(int type)
   return "?????";
 }
 
+static uint32_t _extract_bits(uint32_t num, int low, int size)
+{
+  return (num & (((1 << size) - 1) << low)) >> low;
+}
+
 /****************************************************************************
  * Name: _get_val, set_val, _add_val
  *
@@ -394,7 +400,7 @@ int up_relocateadd(const Elf_Rela *rel, const Elf_Sym *sym,
 
   relotype = ELF_R_TYPE(rel->r_info);
 
-  if (relotype == R_RISCV_RELAX)
+  if (relotype == R_RISCV_RELAX || relotype == R_RISCV_ALIGN)
     {
       /* NOTE: RELAX has no symbol, so just return */
 
@@ -561,12 +567,21 @@ int up_relocateadd(const Elf_Rela *rel, const Elf_Sym *sym,
           uint32_t val = _get_val((uint16_t *)addr) & 0xfe000f80;
 
           /* NOTE: we assume that a compiler adds an immediate value */
+          if (offset != 0 && val == 0)
+            {
+              uint32_t imm12 = _extract_bits(offset, 12, 1) << 31;
+              uint32_t imm10_5 = _extract_bits(offset, 5, 6) << 25;
+              uint32_t imm4_1 = _extract_bits(offset, 1, 4) << 8;
+              uint32_t imm11 = _extract_bits(offset, 11, 1) << 7;
 
-          ASSERT(offset && val);
-
-          binfo("offset for Bx=%ld (0x%lx) (val=0x%08" PRIx32 ") "
-                "already set!\n",
-                offset, offset, val);
+              _add_val((uint16_t *)addr, imm12 | imm10_5 | imm4_1 | imm11);
+            }
+          else
+            {
+              binfo("offset for Bx=%ld (0x%lx) (val=0x%08" PRIx32 ") "
+                    "already set!\n",
+                    offset, offset, val);
+            }
         }
         break;
 
@@ -583,13 +598,22 @@ int up_relocateadd(const Elf_Rela *rel, const Elf_Sym *sym,
           offset = (long)sym->st_value + (long)rel->r_addend - (long)addr;
           uint32_t val = _get_val((uint16_t *)addr) & 0xfffff000;
 
-          ASSERT(offset && val);
-
           /* NOTE: we assume that a compiler adds an immediate value */
+          if (offset != 0 && val == 0)
+            {
+              uint32_t imm20 = _extract_bits(offset, 20, 1) << 31;
+              uint32_t imm10_1 = _extract_bits(offset, 1, 10) << 21;
+              uint32_t imm11 = _extract_bits(offset, 11, 1) << 20;
+              uint32_t imm19_12 = _extract_bits(offset, 12, 8) << 12;
 
-          binfo("offset for JAL=%ld (0x%lx) (val=0x%08" PRIx32 ") "
-                "already set!\n",
-                offset, offset, val);
+              _add_val((uint16_t *)addr, imm20 | imm10_1 | imm11 | imm19_12);
+            }
+          else
+            {
+              binfo("offset for JAL=%ld (0x%lx) (val=0x%08" PRIx32 ") "
+                    "already set!\n",
+                    offset, offset, val);
+            }
         }
         break;
 
@@ -692,8 +716,25 @@ int up_relocateadd(const Elf_Rela *rel, const Elf_Sym *sym,
 
           uint16_t val = (*(uint16_t *)addr) & 0x1ffc;
 
-          binfo("offset for C.J=%ld (0x%lx) (val=0x%04x) already set!\n",
-                offset, offset, val);
+          /* NOTE: we assume that a compiler adds an immediate value */
+          if (offset != 0 && val == 0)
+            {
+              uint16_t imm11 = _extract_bits(offset, 11, 1) << 12;
+              uint16_t imm4 = _extract_bits(offset, 4, 1) << 11;
+              uint16_t imm9_8 = _extract_bits(offset, 8, 2) << 9;
+              uint16_t imm10 = _extract_bits(offset, 10, 1) << 8;
+              uint16_t imm6 = _extract_bits(offset, 6, 1) << 7;
+              uint16_t imm7 = _extract_bits(offset, 7, 1) << 6;
+              uint16_t imm3_1 = _extract_bits(offset, 1, 3) << 3;
+              uint16_t imm5 = _extract_bits(offset, 5, 1) << 2;
+
+              _add_val((uint16_t *)addr, imm11 | imm4 | imm9_8 | imm10 | imm6 | imm7 | imm3_1 | imm5);
+            }
+          else
+            {
+              binfo("offset for C.J=%ld (0x%lx) (val=0x%04x) already set!\n",
+                    offset, offset, val);
+            }
         }
         break;
 
@@ -713,11 +754,21 @@ int up_relocateadd(const Elf_Rela *rel, const Elf_Sym *sym,
           uint16_t val = (*(uint16_t *)addr) & 0x1c7c;
 
           /* NOTE: we assume that a compiler adds an immediate value */
+          if (offset != 0 && val == 0)
+            {
+              uint16_t imm8 = _extract_bits(offset, 8, 1) << 12;
+              uint16_t imm4_3 = _extract_bits(offset, 3, 2) << 10;
+              uint16_t imm7_6 = _extract_bits(offset, 6, 2) << 5;
+              uint16_t imm2_1 = _extract_bits(offset, 1, 2) << 3;
+              uint16_t imm5 = _extract_bits(offset, 5, 1) << 2;
 
-          ASSERT(offset && val);
-
-          binfo("offset for C.Bx=%ld (0x%lx) (val=0x%04x) already set!\n",
-                offset, offset, val);
+              _add_val((uint16_t *)addr, imm8 | imm4_3 | imm7_6 | imm2_1 | imm5);
+            }
+          else
+            {
+              binfo("offset for C.Bx=%ld (0x%lx) (val=0x%04x) already set!\n",
+                    offset, offset, val);
+            }
         }
         break;
       case R_RISCV_32_PCREL:
diff --git a/syscall/syscall.csv b/syscall/syscall.csv
index 6b23fcd604..2cf827df61 100644
--- a/syscall/syscall.csv
+++ b/syscall/syscall.csv
@@ -203,3 +203,15 @@
 "waitid","sys/wait.h","defined(CONFIG_SCHED_WAITPID) && defined(CONFIG_SCHED_HAVE_PARENT)","int","idtype_t","id_t"," FAR siginfo_t *","int"
 "waitpid","sys/wait.h","defined(CONFIG_SCHED_WAITPID)","pid_t","pid_t","FAR int *","int"
 "write","unistd.h","","ssize_t","int","FAR const void *","size_t"
+"webclient_set_defaults","../../apps/include/netutils/webclient.h","","void","FAR struct webclient_context *"
+"webclient_perform","../../apps/include/netutils/webclient.h","","int","FAR struct webclient_context *"
+"webclient_abort","../../apps/include/netutils/webclient.h","","void","FAR struct webclient_context *"
+"mqtt_init","../../apps/netutils/mqttc/MQTT-C/include/mqtt.h","","enum MQTTErrors","struct mqtt_client *, mqtt_pal_socket_handle, uint8_t *, size_t, uint8_t *, size_t, void (*)(void **, struct mqtt_response_publish *)"
+"mqtt_init_reconnect","../../apps/netutils/mqttc/MQTT-C/include/mqtt.h","","void","struct mqtt_client *, void (*)(struct mqtt_client *, void **), void *, void (*)(void **, struct mqtt_response_publish *)"
+"mqtt_connect","../../apps/netutils/mqttc/MQTT-C/include/mqtt.h","","enum MQTTErrors","struct mqtt_client *, const char *, const char *, const void *, size_t, const char*, const char*, uint8_t, uint16_t"
+"mqtt_publish","../../apps/netutils/mqttc/MQTT-C/include/mqtt.h","","enum MQTTErrors","struct mqtt_client *, const char *, const void *, size_t, uint8_t"
+"mqtt_subscribe","../../apps/netutils/mqttc/MQTT-C/include/mqtt.h","","enum MQTTErrors","struct mqtt_client *, const char *, int"
+"mqtt_unsubscribe","../../apps/netutils/mqttc/MQTT-C/include/mqtt.h","","enum MQTTErrors","struct mqtt_client *, const char *"
+"mqtt_disconnect","../../apps/netutils/mqttc/MQTT-C/include/mqtt.h","","enum MQTTErrors","struct mqtt_client *"
+"mqtt_reconnect","../../apps/netutils/mqttc/MQTT-C/include/mqtt.h","","enum MQTTErrors","struct mqtt_client *"
+"sslutil_webclient_tlsops","../arch/risc-v/src/common/espressif/sslutil.h","","struct webclient_tls_ops *","void"
